R/chop.R

```diff
@@ -90,188 +90,253 @@ chop <- function(data, cols) {
+  ellipsis::check_dots_empty()
+
+  if (!is.data.frame(x)) {
+    abort("`x` must be a data frame.")
+  }
+  if (!is_list(ptype)) {
+    abort("`ptype` must be a list.")
+  }
+  if (!is_bool(keep_empty)) {
+    abort("`keep_empty` must be a single `TRUE` or `FALSE`.")
+  }
+
   size <- vec_size(x)
 
+  # Avoid any data frame subclass method dispatch
+  x <- new_data_frame(x, n = size)
+
+  width <- length(x)
+  names <- names(x)
+
   seq_len_width <- seq_len(width)
   seq_len_size <- seq_len(size)
 
-  sizes <- rep_len(NA_integer_, size)
+  if (width == 0L) {
+    # Algorithm requires >=1 columns
+    out <- list(loc = seq_len_size, val = x)
+    out <- new_data_frame(out, n = size)
+    return(out)
+  }
+
+  x_is_list <- map_lgl(x, vec_is_list)
+
+  x_sizes <- vector("list", length = width)
+  x_nulls <- vector("list", length = width)
 
-  # Gather the common size of each row.
-  # Effectively equivalent to creating a `[vec_size(x), length(x)]` matrix,
-  # taking the size of each individual element of `x`, and then taking the
-  # common size of each row.
-  # `NULL` elements are ignored in the size calculation by treating their
-  # size as `NA` and then deferring to the size of any other element in the row.
-  # If only `NULL` values are in the row, the `NA` size is finalised to `0`.
   for (i in seq_len_width) {
     col <- x[[i]]
+    col_is_list <- x_is_list[[i]]
 
-    for (j in seq_len_size) {
-      # TODO: col[[j]] -> vec_slice2(col, j)
-      elt <- col[[j]]
+    if (!col_is_list) {
+      # Optimize rare non list-cols
+      x_sizes[[i]] <- vec_rep(1L, size)
+      x_nulls[[i]] <- vec_rep(FALSE, size)
+      next
+    }
 
-      old_size <- sizes[[j]]
-      new_size <- tidyr_size(elt)
+    info <- unchop_col_info(col, size, keep_empty)
 
-      sizes[[j]] <- tidyr_size2(old_size, new_size)
-    }
+    x[[i]] <- info$col
+    x_sizes[[i]] <- info$sizes
+    x_nulls[[i]] <- info$null
   }
 
-  sizes <- map_int(sizes, tidyr_size_finalise)
+  sizes <- reduce(x_sizes, unchop_sizes2)
 
-  has_ptype <- !is.null(ptype)
-  if (has_ptype && !is.data.frame(ptype)) {
-    abort("`ptype` must be a data frame")
-  }
+  info <- unchop_finalize(x, sizes, x_nulls, keep_empty)
+  x <- info$x
+  sizes <- info$sizes
 
-  # Initialize `cols` with ptypes to retain types when `x` has 0 size
-  cols <- map(x, df_unchop_ptype)
-  pieces <- vector("list", size)
+  out_loc <- vec_rep_each(seq_len_size, sizes)
 
-  names <- names(x)
-  names(cols) <- names
+  out_size <- sum(sizes)
+  out_cols <- vector("list", length = width)
 
   for (i in seq_len_width) {
     col <- x[[i]]
+    col_is_list <- x_is_list[[i]]
 
-    for (j in seq_len_size) {
-      # TODO: col[[j]] -> vec_slice2(col, j)
-      elt <- col[[j]]
-      size <- sizes[[j]]
-
-      # Recycle each row element to the common size of that row
-      pieces[[j]] <- tidyr_recycle(elt, size)
+    if (!col_is_list) {
+      out_cols[[i]] <- vec_slice(col, out_loc)
```

Comment:
Nice optimization also pointed out by @mgirlich that while unchopping non-list columns is kind of a silly thing for the user to do, it can be done a lot faster by just slicing the column. Previously I think we essentially converted it to a list, recycled each element, and then converted it back to a vector.

This will actually help performance with https://github.com/tidyverse/tidyr/issues/1112. The user is trying to `unnest()` a data frame column (really they should use `unpack()`, but `unnest()` should work too). The `unchop()` step of that will have them unchopping a df-col, which will be faster because of this optimization.